File: frontend/src/features/workspace/types/template.ts
/**
 * 模板系统类型定义
 *
 * Phase 4: 模板市场相关类型
 */

/**
 * 变量定义
 */
export interface VariableSchema {
  name: string
  var_type: 'string' | 'number' | 'boolean' | 'datasource' | 'list'
  description: string
  default?: unknown
  required: boolean
  enum_values?: unknown[]
}

/**
 * 模板响应
 */
export interface TemplateResponse {
  template_id: string
  name: string
  description: string
  category: string | null
  author: string | null
  tags: string[]
  usage_count: number
  preview_image: string | null
  version: string

  // 变量定义
  variables: Record<string, VariableSchema>

  // 步骤概要
  step_count: number
  step_types: string[]

  created_at: string
  updated_at: string
}

/**
 * 模板列表响应
 */
export interface TemplateListResponse {
  templates: TemplateResponse[]
  total: number
  categories: Record<string, number>
}

/**
 * 模板列表查询参数
 */
export interface TemplateListQuery {
  category?: string | null
  tags?: string[]
  search?: string
  sort_by?: 'usage_count' | 'created_at' | 'name'
  limit?: number
  offset?: number
}

/**
 * 创建模板请求
 */
export interface CreateTemplateRequest {
  workflow_id: string
  category: string
  author?: string
  preview_image?: string
  tags?: string[]
}

/**
 * 实例化模板请求
 */
export interface InstantiateRequest {
  variable_values: Record<string, unknown>
  new_name?: string
}

/**
 * 实例化响应
 */
export interface InstantiateResponse {
  workflow_id: string
  name: string
  status: string
  template_source_id: string
}

/**
 * 变量校验响应
 */
export interface ValidateVariablesResponse {
  valid: boolean
  errors: string[]
}

/**
 * 分类统计
 */
export interface CategoryStats {
  category: string
  label: string
  count: number
}

/**
 * 模板市场统计
 */
export interface TemplateStatsResponse {
  total: number
  categories: CategoryStats[]
}

/**
 * 导入模板请求
 */
export interface ImportTemplateRequest {
  template_data: Record<string, unknown>
  author?: string
}

/**
 * 模板分类枚举
 */
export const TEMPLATE_CATEGORIES = {
  data_analysis: '数据分析',
  content_research: '内容研究',
  competitive: '竞品分析',
  social_monitoring: '社交监控',
  report_generation: '报告生成',
  custom: '自定义',
} as const

export type TemplateCategoryKey = keyof typeof TEMPLATE_CATEGORIES

File: api/schemas/template.py
"""
模板 API Schemas

Phase 4: 模板市场相关的请求/响应模型
"""

from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class VariableSchema(BaseModel):
    """变量定义 Schema"""
    name: str
    var_type: str  # string | number | boolean | datasource | list
    description: str = ""
    default: Optional[Any] = None
    required: bool = True
    enum_values: Optional[List[Any]] = None


class TemplateListQuery(BaseModel):
    """模板列表查询参数"""
    category: Optional[str] = Field(None, description="按分类筛选")
    tags: Optional[List[str]] = Field(None, description="按标签筛选")
    search: Optional[str] = Field(None, description="搜索关键词")
    sort_by: str = Field("usage_count", description="排序方式: usage_count | created_at | name")
    limit: int = Field(20, ge=1, le=100, description="返回数量")
    offset: int = Field(0, ge=0, description="偏移量")


class TemplateResponse(BaseModel):
    """模板响应"""
    template_id: str
    name: str
    description: str
    category: Optional[str] = None
    author: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    usage_count: int = 0
    preview_image: Optional[str] = None
    version: str = "1.0.0"

    # 变量定义
    variables: Dict[str, VariableSchema] = Field(default_factory=dict)

    # 步骤概要
    step_count: int = 0
    step_types: List[str] = Field(default_factory=list)

    created_at: str
    updated_at: str


class TemplateListResponse(BaseModel):
    """模板列表响应"""
    templates: List[TemplateResponse]
    total: int
    categories: Dict[str, int] = Field(default_factory=dict)


class CreateTemplateRequest(BaseModel):
    """创建模板请求"""
    workflow_id: str = Field(..., description="源工作流 ID")
    category: str = Field("custom", description="模板分类")
    author: str = Field("anonymous", description="作者")
    preview_image: Optional[str] = Field(None, description="预览图 URL")
    tags: Optional[List[str]] = Field(None, description="标签")


class InstantiateRequest(BaseModel):
    """实例化模板请求"""
    variable_values: Dict[str, Any] = Field(default_factory=dict, description="变量值")
    new_name: Optional[str] = Field(None, description="新工作流名称")


class InstantiateResponse(BaseModel):
    """实例化响应"""
    workflow_id: str
    name: str
    status: str
    template_source_id: str


class ValidateVariablesRequest(BaseModel):
    """变量校验请求"""
    variable_values: Dict[str, Any]


class ValidateVariablesResponse(BaseModel):
    """变量校验响应"""
    valid: bool
    errors: List[str] = Field(default_factory=list)


class ImportTemplateRequest(BaseModel):
    """导入模板请求"""
    template_data: Dict[str, Any] = Field(..., description="模板 JSON 数据")
    author: str = Field("imported", description="导入后的作者标记")


class CategoryStats(BaseModel):
    """分类统计"""
    category: str
    label: str
    count: int


class TemplateStatsResponse(BaseModel):
    """模板市场统计响应"""
    total: int
    categories: List[CategoryStats]
